### PR TITLE
[ENV] Allows for environment settings to be loaded dynamically via a YAML settings file

### PR DESCRIPTION
**Pull request by eliasjpr**
### Description of the Change

Allows for environment settings to be loaded dynamically via a YAML settings file.

- Default environment path: `{project root}/config/environments`

### Usage
```crystal
Amber.env = :production
Amber.env = :test
Amber.env = :development

# Changing environment path
Amber.environment_path = "Some/Other/Path"

# Gets current environment settings
Amber.settings

# Get logger
Amber.logger

```

### Benefits

- Adds documentation to `config/application.cr` for each setting
- Conveniently easy to switch environments
- Easy to add more settings variables
- Easy to test
- All server settings are referenced via `Amber.settings` **Updated**
- All Amber logging is now done thru `Amber.logger` **Updated**
- Parallelizes Travis tests by component
- Command Exec uses `config/application.cr` since this is the main entry-point file 
- Adds helper methods `Amber.secret_key`

### Possible Drawbacks
- Cannot embed Dynamic code in yaml settings file.
- Must use `config/application.cr` for any dynamic settings including loading environment variables.

### CODE REVIEW

File: src/amber/cli/templates/app/config/application.cr.ecr [hunk 1]
```diff
@@ -1,16 +1,115 @@
...
 
-# NOTE: Settings should all be in config/environments/env.yml.
-# Anything here will overwrite all environments.
-Amber::Server.configure do |setting|
-  # Server options
-  # setting.name = "<%= display_name %> web application."
-  # setting.port = 80 # Port you wish your app to run
-  # setting.log = ::Logger.new(STDOUT)
-  # setting.log.level = ::Logger::INFO
+# About Application.cr File
```
Comment (drujensen) [inline src/amber/cli/templates/app/config/application.cr.ecr]: 🥇 excellent documentation!

File: src/amber/cli/templates/app/config/application.cr.ecr [hunk 2]
```diff
@@ -1,16 +1,116 @@
...
+# Amber::Server boot up.
+#
+# > We recommend to not modify the order of the require since the order will
+# affect the behavior of the application.
+#
+# With `Amber::Server.configure` block you can redefine the Server configuration
+# settings and use ENVIRONMENT variables and/or values evaluated at runtime.
+#
+#
+Amber::Server.configure do |settings|
```
Comment (elorest) [inline src/amber/cli/templates/app/config/application.cr.ecr]: I'm concerned that this will encourage them to configure their environments here instead of using `config/environments/env.yml`. 

Perhaps we should only show a couple examples that relate to ENV vars and show how to optionally override.

Most settings should be set in the env files. Not here!

File: src/amber/cli/templates/app/config/application.cr.ecr [hunk 3]
```diff
@@ -1,16 +1,116 @@
...
+  # .amber_secret_key in this order.
+  #
+  # settings.secret_key_base= ENV["SECRET_KEY_BASE"]
+  #
+  #
+  # Host: is the application server host address or ip address. Useful for when 
+  # deploying Amber to a PAAS and likely the assigned server IP is either 
+  # known or unknown. Defaults to an environment variable HOST
+  #
+  # settings.host = ENV["HOST"]
```
Comment (elorest) [inline src/amber/cli/templates/app/config/application.cr.ecr]: ENV overrides should probably be setup to over ride if ENV exists but otherwise use the default settings from env.ymls.

EXAMPLE: 

```cr
settings.host = ENV["HOST"] if ENV["HOST"]?
```

This way if HOST is empty it will keep the existing value instead of setting it to nil.

File: src/amber/cli/templates/app/config/environments/development.yml.ecr [hunk 4]
```diff
@@ -18,9 +16,11 @@ database_url: postgres://postgres:@localhost:5432/<%= database_name_base %>_deve
...
-  store: :signed_cookie
-  expires: 0
-secrets:
-  description: Store your development secrets credentials and settings here.
+session: {
+  key: amber.session,
+  store: signed_cookie,
+  expires: 0,
+}
+secrets: {
```
Comment (elorest) [inline src/amber/cli/templates/app/config/environments/development.yml.ecr]: Why does this have curly brackets? Shouldn't it just be a tab to turn into has with yaml?

File: src/amber/cli/templates/app/config/environments/production.yml.ecr [hunk 5]
```diff
@@ -1,14 +1,12 @@
 secret_key_base: <%= SecureRandom.urlsafe_base64(32) %>
-port: <%= (ENV["PORT"] ||= "80").to_i %>
+port: 8080
 name: <%= @name %>
-log: "::Logger.new(STDOUT)"
```
Comment (elorest) [inline src/amber/cli/templates/app/config/environments/production.yml.ecr]: How do we configure loggers per environment this way?

File: src/amber/scripts/environment_loader.cr [hunk 6]
```diff
@@ -1,75 +0,0 @@
...
-        "env: #{environment}"
-      end
-
-settings = YAML.parse(yml)
-
-str = String.build do |s|
-  # Most of this logic can be cleaned up by just requiring environment files to contain valid params.
-  # For now I have this here so that tests can still pass without having to load env files although they should.
-  # This is a transistion.
-  s.puts %(@name = "#{settings["name"]? || "Amber_App"}")
```
Comment (elorest) [inline src/amber/scripts/environment_loader.cr]: So much red. I agree this is for the best but it makes me sad.

File: src/amber/server/server.cr [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    def self.start
+      instance.run
+    end
+
+    # Configure should probably be deprecated in favor of settings.
+    def self.configure
+      with self yield settings
+    end
+
+    def self.settings
```
Comment (elorest) [inline src/amber/server/server.cr]: So this is all back here now? I'm fine with it. Why did you move it out to `configuration.cr` just to move it back here though?

File: src/amber/server/settings.cr [hunk 8]
```diff
@@ -1,26 +1,63 @@
...
-    property host : String
-    property ssl_key_file : String? = nil
-    property ssl_cert_file : String? = nil
-    property redis_url = ""
+    property log : ::Logger = Logger.new(STDOUT)
+
+    property colorize_logging : Bool = true
     property database_url = ""
-    property session : Hash(Symbol, Symbol | String | Int32)
+    property host : String = "localhost"
```
Comment (elorest) [inline src/amber/server/settings.cr]: We shouldn't be defining defaults in both property definitions and yaml mappings. I'd vote for putting all defaults in yaml mappings.

File: src/amber/server/settings.cr [hunk 9]
```diff
@@ -1,26 +1,63 @@
...
-    property color : Bool = true
-    property process_count : Int32
-    property secret_key_base : String
-    property port : Int32
-    property name : String
-    property host : String
-    property ssl_key_file : String? = nil
-    property ssl_cert_file : String? = nil
-    property redis_url = ""
+    property log : ::Logger = Logger.new(STDOUT)
```
Comment (elorest) [inline src/amber/server/settings.cr]: I don't think we should have properties of settings that can't be set through environment.yms's.

File: src/amber/websockets/adapters/redis.cr [hunk 10]
```diff
@@ -10,8 +10,8 @@ module Amber::WebSockets::Adapters
 
     # Establish subscribe and publish connections to Redis
     def initialize
-      @subscriber = Redis.new(url: Amber::Server.redis_url)
-      @publisher = Redis.new(url: Amber::Server.redis_url)
+      @subscriber = Redis.new(url: Amber::Server.settings.redis_url)
```
Comment (elorest) [inline src/amber/websockets/adapters/redis.cr]: Weren't you planning on moving all of these straight to Amber like `Amber.settings.redis_url`.

File: src/amber/cli/templates/app/config/environments/production.yml.ecr (hunk 5 above)
Comment (eliasjpr) [reply src/amber/cli/templates/app/config/environments/production.yml.ecr]: Honestly defining a Logger should not be part of a setting. I am considering to move the logger to `Amber.logger`

File: src/amber/server/server.cr (hunk 7 above)
Comment (eliasjpr) [reply src/amber/server/server.cr]: Removing this in favor of `Amber.settings`

File: src/amber/server/settings.cr (hunk 9 above)
Comment (eliasjpr) [reply src/amber/server/settings.cr]: I agree with you. But for now I am going to leave as is for now until we can come up with a better soloution. WDYT?

File: src/amber/server/settings.cr (hunk 8 above)
Comment (eliasjpr) [reply src/amber/server/settings.cr]: Completely agree.

File: src/amber/websockets/adapters/redis.cr (hunk 10 above)
Comment (eliasjpr) [reply src/amber/websockets/adapters/redis.cr]: Just did 😃

File: src/amber/cli/templates/app/config/application.cr.ecr (hunk 2 above)
Comment (eliasjpr) [reply src/amber/cli/templates/app/config/application.cr.ecr]: I will add a better note that the YAMLs are first class citizen and only use this if completely necessary.

File: src/amber/scripts/environment_loader.cr (hunk 6 above)
Comment (eliasjpr) [reply src/amber/scripts/environment_loader.cr]: Don't be, most of the inspiration came from your version 😄

File: src/amber/cli/templates/app/config/environments/development.yml.ecr (hunk 4 above)
Comment (elorest) [reply src/amber/cli/templates/app/config/environments/development.yml.ecr]: In my tests with yaml it seems to work just fine without them...

```yaml
secrets:
  name: "isaac"
  age: 34
  hash: true
```

File: src/amber/cli/templates/app/config/environments/production.yml.ecr (hunk 5 above)
Comment (elorest) [reply src/amber/cli/templates/app/config/environments/production.yml.ecr]: Would it be possible to completely configure the logger through the pipeline rather than here?

`plug Amber::Pipe::FastLogger.new(stdout: Amber.env.development?)`

**Comment by elorest** [comment]
Thanks

**Comment by eliasjpr** [comment]
@amberframework/core-team made some updates based on @elorest suggestions. The updates includes

- Amber.logger  - Gets the logger from the settings
- Amber.settings - Gets `current` environment settings
- Removes curly braces from YAML - it turns out that it just needed a break-line between the session and secrets 
- Travis.yml - Reduce the number of parallel test run and moves the build_spec test to the root of the root. This allows to run the latter in isolation.

What I did not do:

**Application.cr** @elorest brought up a good point around the environment variables in the application.cr file. He suggested to use `settings.host = ENV["HOST"] if ENV["HOST"]?`

> I decided to not add the if statement because the settings in this file are by default commented out and it explicitly states that if used it will override the settings value from the YAML for the environment loaded. It is up to the user to use define and use the environment variables accordingly.

**Comment by faustinoaq** [comment]
Is this PR the first step to achieve #346?

**Comment by elorest** [comment]
@eliasjpr 

Thanks for working on this over Thanksgiving.

I'd like to point out that my suggestion for environment variables is bigger than just functionality. If someone were to uncomment host and port. But ever not set one environment variable it will exception on index out of bounds and crash. This behavior will be unexpected since we provided them this code to uncomment. 

IMO the expected behavior is "Use overwrite default settings if ENV VAR exists", not "use ENV VAR or Crash if it doesn't exist."

```cr
settings.host = ENV["HOST"]
settings.port = ENV["PORT"]
```
We should provide a few example of using ENV VAR only if it exists. Otherwise we're going to have to explain this a lot of times in gitter.

WDYT?

**Comment by eliasjpr** [comment]
@elorest IMHO think is a good thing that it would raise an exception if the **ENV** variable was never defined at least I know as a developer I forgot to set something up. Not sure if this is a blocking issue for this PR. I would like for use to raise a discussion around that in the core team channel and get feedback from everyone, and address the outcome in a separate PR.

**Comment by elorest** [comment]
Ok. To me it's not that it raises an exception that has me concerned, as it will work in development if they define an ENV VAR. As soon as it's not set though on a different server it will error. This could result in an error that's hard to track down. 

Obviously it's fine if the users write code like this. I'm just concerned that we seem to be encouraging it. 

Within Rails and Phoenix every ENV override has default values that are used instead if they're not set.

If a PORT isn't set it goes to 3000, if an RAILS_ENV isn't set it goes to development, if DATABASE_URL isn't set it uses settings from database.yml.

I don't think that it's clear to users that uncommenting the code in `application.cr` will work differently.
